Expose device related test classes for reusing

EikanWang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122953

This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `@skipIfXXX`, `@onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
[test_tensor_creation_ops.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-bbf4fd94ba055c8e116d76a113c0ba3624c730bfefa0f8784a5aa5ba3b67397e)
```python
@unMarkDynamoStrictTest
class TestTensorCreation(TestCase):
  ...

class Namespace:
    class TestTensorCreationWrapper(TestTensorCreation):
        pass

instantiate_device_type_tests(TestTensorCreation, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

[test_tensor_creation_ops_xpu.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-230214b93c18991251c2cfa93df651ada20607e7a6f50aa3903238bc564b2360)
```python
class TestTensorCreationXPU(TestCase):
    pass
...

copy_tests(
    TestTensorCreationXPU,
    Namespace.TestTensorCreationWrapper,
    applicable_list=["test_empty_strided"],
)
instantiate_device_type_tests(TestTensorCreationXPU, globals(), only_for="xpu")
```

EikanWang committed:
Expose test classes for test cases reusing

[ghstack-poisoned]

EikanWang committed:
Update on "[WIP] Expose test classes for test cases reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by inheriting the wrapper class and then run the applicable test cases.


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
test_ops.py
```python
unMarkDynamoStrictTest
class TestCommon(TestCase):
  ...

class Namespace:
    class TestCommonWrapper(TestCommon):
        pass

instantiate_device_type_tests(TestCommon, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

test_ops_xpu.py
```python
class TestXPUOpsTest(TestCase):
  pass

...
instantiate_copy_device_type_tests_from(TestXPUOpsTest, Namespace.TestCommonWrapper, applicable_list=["test_compare_cpu"])
instantiate_device_type_tests(TestXPUOpsTest, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
test_ops.py
```python
unMarkDynamoStrictTest
class TestCommon(TestCase):
  ...

class Namespace:
    class TestCommonWrapper(TestCommon):
        pass

instantiate_device_type_tests(TestCommon, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

test_ops_xpu.py
```python
class TestXPUOpsTest(TestCase):
  pass

...
instantiate_copy_device_type_tests_from(TestXPUOpsTest, Namespace.TestCommonWrapper, applicable_list=["test_compare_cpu"])
instantiate_device_type_tests(TestXPUOpsTest, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
test_ops.py
```python
unMarkDynamoStrictTest
class TestCommon(TestCase):
  ...

class Namespace:
    class TestCommonWrapper(TestCommon):
        pass

instantiate_device_type_tests(TestCommon, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

test_ops_xpu.py
```python
class TestXPUOpsTest(TestCase):
  pass

...
instantiate_copy_device_type_tests_from(TestXPUOpsTest, Namespace.TestCommonWrapper, applicable_list=["test_compare_cpu"])
instantiate_device_type_tests(TestXPUOpsTest, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
test_ops.py
```python
unMarkDynamoStrictTest
class TestCommon(TestCase):
  ...

class Namespace:
    class TestCommonWrapper(TestCommon):
        pass

instantiate_device_type_tests(TestCommon, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

test_ops_xpu.py
```python
class TestXPUOpsTest(TestCase):
  pass

...
instantiate_device_type_tests_copy(TestXPUOpsTest, Namespace.TestCommonWrapper, applicable_list=["test_compare_cpu"])
instantiate_device_type_tests(TestXPUOpsTest, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
test_ops.py
```python
unMarkDynamoStrictTest
class TestCommon(TestCase):
  ...

class Namespace:
    class TestCommonWrapper(TestCommon):
        pass

instantiate_device_type_tests(TestCommon, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

test_ops_xpu.py
```python
class TestXPUOpsTest(TestCase):
  pass

...
instantiate_device_type_tests_copy(TestXPUOpsTest, Namespace.TestCommonWrapper, applicable_list=["test_compare_cpu"])
instantiate_device_type_tests(TestXPUOpsTest, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
test_ops.py
```python
unMarkDynamoStrictTest
class TestCommon(TestCase):
  ...

class Namespace:
    class TestCommonWrapper(TestCommon):
        pass

instantiate_device_type_tests(TestCommon, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

test_ops_xpu.py
```python
class TestXPUOpsTest(TestCase):
  pass

...
copy_tests(TestXPUOpsTest, Namespace.TestCommonWrapper, applicable_list=["test_compare_cpu"])
instantiate_device_type_tests(TestXPUOpsTest, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
test_ops.py
```python
unMarkDynamoStrictTest
class TestCommon(TestCase):
  ...

class Namespace:
    class TestCommonWrapper(TestCommon):
        pass

instantiate_device_type_tests(TestCommon, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

test_ops_xpu.py
```python
class TestXPUOpsTest(TestCase):
  pass

...
copy_tests(TestXPUOpsTest, Namespace.TestCommonWrapper, applicable_list=["test_compare_cpu"])
instantiate_device_type_tests(TestXPUOpsTest, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
[test_tensor_creation_ops.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-bbf4fd94ba055c8e116d76a113c0ba3624c730bfefa0f8784a5aa5ba3b67397e)
```python
unMarkDynamoStrictTest
class TestTensorCreation(TestCase):
  ...

class Namespace:
    class TestTensorCreationWrapper(TestTensorCreation):
        pass

instantiate_device_type_tests(TestTensorCreation, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

[test_tensor_creation_ops_xpu.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-230214b93c18991251c2cfa93df651ada20607e7a6f50aa3903238bc564b2360)
```python
class TestTensorCreationXPU(TestCase):
    pass
...

copy_tests(
    TestTensorCreationXPU,
    Namespace.TestTensorCreationWrapper,
    applicable_list=["test_empty_strided"],
)
instantiate_device_type_tests(TestTensorCreationXPU, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
[test_tensor_creation_ops.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-bbf4fd94ba055c8e116d76a113c0ba3624c730bfefa0f8784a5aa5ba3b67397e)
```python
unMarkDynamoStrictTest
class TestTensorCreation(TestCase):
  ...

class Namespace:
    class TestTensorCreationWrapper(TestTensorCreation):
        pass

instantiate_device_type_tests(TestTensorCreation, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

[test_tensor_creation_ops_xpu.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-230214b93c18991251c2cfa93df651ada20607e7a6f50aa3903238bc564b2360)
```python
class TestTensorCreationXPU(TestCase):
    pass
...

copy_tests(
    TestTensorCreationXPU,
    Namespace.TestTensorCreationWrapper,
    applicable_list=["test_empty_strided"],
)
instantiate_device_type_tests(TestTensorCreationXPU, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
[test_tensor_creation_ops.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-bbf4fd94ba055c8e116d76a113c0ba3624c730bfefa0f8784a5aa5ba3b67397e)
```python
unMarkDynamoStrictTest
class TestTensorCreation(TestCase):
  ...

class Namespace:
    class TestTensorCreationWrapper(TestTensorCreation):
        pass

instantiate_device_type_tests(TestTensorCreation, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

[test_tensor_creation_ops_xpu.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-230214b93c18991251c2cfa93df651ada20607e7a6f50aa3903238bc564b2360)
```python
class TestTensorCreationXPU(TestCase):
    pass
...

copy_tests(
    TestTensorCreationXPU,
    Namespace.TestTensorCreationWrapper,
    applicable_list=["test_empty_strided"],
)
instantiate_device_type_tests(TestTensorCreationXPU, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
[test_tensor_creation_ops.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-bbf4fd94ba055c8e116d76a113c0ba3624c730bfefa0f8784a5aa5ba3b67397e)
```python
unMarkDynamoStrictTest
class TestTensorCreation(TestCase):
  ...

class Namespace:
    class TestTensorCreationWrapper(TestTensorCreation):
        pass

instantiate_device_type_tests(TestTensorCreation, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

[test_tensor_creation_ops_xpu.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-230214b93c18991251c2cfa93df651ada20607e7a6f50aa3903238bc564b2360)
```python
class TestTensorCreationXPU(TestCase):
    pass
...

copy_tests(
    TestTensorCreationXPU,
    Namespace.TestTensorCreationWrapper,
    applicable_list=["test_empty_strided"],
)
instantiate_device_type_tests(TestTensorCreationXPU, globals(), only_for="xpu")
```


[ghstack-poisoned]

albanD commented:
cc @seemethere who is the right person to review test infrastructure changes?

malfet commented on test/run_test.py:
```diff
@@ -182,9 +182,7 @@ def __contains__(self, item):
-XPU_TEST = [
-    "test_xpu",
-]
+XPU_TEST = ["test_xpu", "xpu/test_tensor_creation_ops_xpu"]
```
Nit (can we keep a multiline definition here)
```suggestion
XPU_TEST = [
        "test_xpu",
        "xpu/test_tensor_creation_ops_xpu",
        ]
```

EikanWang replied:
`lintrunner` will automatically format the style. It will transfer the multiline to a single line if the number of characters of a line does not exceed the threshold. I will try to keep the multiline and check for any lint issue.

malfet reviewed (changes requested):
Please do not try to patch a syspath, but rather move shared code to torch.testing._internal

EikanWang commented:
I reviewed the design again. For XPU, we intend to reuse the particular entire test class and customize its test cases on demand.

Take `test/test_tensor_creation_ops.py` as an example. It contains three test classes: `TestTensorCreation`, `TestRandomTensorCreation`, and `TestLikeTensorCreation`. For XPU, we just reuse partial test cases of `TestTensorCreation` at this moment. With the XPU aten operation implementation moving forward, we will need to reuse more test cases. 

If we add the test cases to be reused to `torch.testing._internal`, I'm concerned it will bloat the code of `torch.testing._internal` dramatically. And the situation will be more intensive when we need to reuse more test cases in other test files like `test_ops.py,` and `test_pooling.py`. Do you think that will be an issue? @malfet 

Meanwhile, the motivation to add the `test` folder to `sys.path` is to import the test classes that may come from different test files. We can isolate the `sys.path` change for test class import only through python context. May I know your detailed concerns? @malfet

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
[test_tensor_creation_ops.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-bbf4fd94ba055c8e116d76a113c0ba3624c730bfefa0f8784a5aa5ba3b67397e)
```python
unMarkDynamoStrictTest
class TestTensorCreation(TestCase):
  ...

class Namespace:
    class TestTensorCreationWrapper(TestTensorCreation):
        pass

instantiate_device_type_tests(TestTensorCreation, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

[test_tensor_creation_ops_xpu.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-230214b93c18991251c2cfa93df651ada20607e7a6f50aa3903238bc564b2360)
```python
class TestTensorCreationXPU(TestCase):
    pass
...

copy_tests(
    TestTensorCreationXPU,
    Namespace.TestTensorCreationWrapper,
    applicable_list=["test_empty_strided"],
)
instantiate_device_type_tests(TestTensorCreationXPU, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang committed:
Update on "Expose device related test classes for reusing"


This PR intends to enable a new path for device-related test cases reused for new hardware. Currently, there are two paths to utilize the device-related test case for new hardware.

- Approach 1: Refine the test cases by adding new device-specific decorators like `skipIfXXX`, `onlyXXX`
- Approach 2: Port the test cases through copy-paste for newly added hardware

In terms of Approach 1, the test files need to be intrusively changed by adding the new decorators. For newly added hardware, the torch functionalities support is step-by-step. This means that the test files have to be carefully maintained for the newly added hardware w/ frequent PR submissions.

Regarding Approach 2, test cases poring for new hardware through copy-past also require massive maintenance efforts because the PyTorch test cases may be refined anytime by adding or removing some test cases. So, it is barely able to catch up with the latest PyTorch test cases.

To address the pros of both Approach 1 and Approach 2, this PR intends to refine Approach 2 by inheriting the device-related test case class and then customizing the cases for new hardware. The process could be as follows.

- Create a wrapper for each device-related test case class as a nested class, and the wrapper will inherit from the device-related test case class.
  - This prevents test runners from picking it up and running it.
  - This allows other test classes to inherit from the wrapper and then reuse the test cases.
    `instantiate_device_type_tests` will instantiate the device-related test case class and change the class name and test case name to be device-specific. For example, `instantiate_device_type_tests(TestCommon, globals())` will change `TestCommon` to `TestCommonCPU` and `TestCommonCUDA`, while the `device` parameter of test cases will be instantiated to `torch.device('cpu')` or `torch.device('cuda')`

- Create a new hardware-specific test case class by automatically copying the wrapper class and then running the applicable test cases.

For example, the implementation of the new test for xpu could be as follows.
[test_tensor_creation_ops.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-bbf4fd94ba055c8e116d76a113c0ba3624c730bfefa0f8784a5aa5ba3b67397e)
```python
unMarkDynamoStrictTest
class TestTensorCreation(TestCase):
  ...

class Namespace:
    class TestTensorCreationWrapper(TestTensorCreation):
        pass

instantiate_device_type_tests(TestTensorCreation, globals())
...

if __name__ == "__main__":
    TestCase._default_dtype_check_enabled = True
    run_tests()
```

[test_tensor_creation_ops_xpu.py](https://github.com/pytorch/pytorch/pull/122953/files#diff-230214b93c18991251c2cfa93df651ada20607e7a6f50aa3903238bc564b2360)
```python
class TestTensorCreationXPU(TestCase):
    pass
...

copy_tests(
    TestTensorCreationXPU,
    Namespace.TestTensorCreationWrapper,
    applicable_list=["test_empty_strided"],
)
instantiate_device_type_tests(TestTensorCreationXPU, globals(), only_for="xpu")
```


[ghstack-poisoned]

EikanWang commented:
I refined the PR. Compared to the last commit, the latest PR isolates all the XPU-specific changes and does not need to apply any changes to the test files that will be reused for XPU. @malfet

EikanWang committed:
Update

[ghstack-poisoned]

EikanWang commented:
@pytorchbot rebase

EikanWang committed:
Update

[ghstack-poisoned]

malfet reviewed (approved):
It's XPU specific, so whatever goes, but it feels very wrong to me, as it will create more an more obstacles

EikanWang commented:
> it will create more an more obstacles

@malfet , agree with you. I will explore other solutions for test case reusing.
